contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+};
+
+class USDScene::Reader : public USDScene::IO
+{
+	public:
+		Reader( const std::string &fileName ) : IO( fileName )
+		{
+			m_usdStage = pxr::UsdStage::Open( fileName );
+			m_timeCodesPerSecond = m_usdStage->GetTimeCodesPerSecond();
+			m_rootPrim = m_usdStage->GetPseudoRoot();
+		}
+
+		pxr::UsdPrim root() const override
+		{
+			return m_rootPrim;
+		}
+
+		pxr::UsdTimeCode getTime( double timeSeconds ) const override
+		{
+			return timeSeconds * m_timeCodesPerSecond;
+		}
+
+	private:
+		pxr::UsdStageRefPtr m_usdStage;
+		pxr::UsdPrim m_rootPrim;
+
+		double m_timeCodesPerSecond;
+};
+
+class USDScene::Writer : public USDScene::IO
+{
+	public:
+		Writer( const std::string &fileName ) : IO( fileName )
+		{
+			m_usdStage = pxr::UsdStage::CreateNew( "gaffer" );
+			m_rootPrim = m_usdStage->GetPseudoRoot();
+		}
+
+		pxr::UsdPrim root() const override
+		{
+			return m_rootPrim;
+		}
+
+		pxr::UsdTimeCode getTime( double timeSeconds ) const override
+		{
+			return 0.0;
+		}
+
+	private:
+		pxr::UsdStageRefPtr m_usdStage;
+		pxr::UsdPrim m_rootPrim;
+};
+
+USDScene::USDScene( const std::string &path, IndexedIO::OpenMode &mode )
+{
+	switch( mode )
+	{
+		case IndexedIO::Read :
+			m_root = new Reader( path );
+			m_prim = m_root->root();
+			break;
+		case IndexedIO::Write :
+			m_root = new Writer( path );
+			m_prim = m_root->root();
+			break;
+		default:
+			throw Exception( " Unsupported OpenMode " );
+	}
+}
+
+USDScene::USDScene( IOPtr root, pxr::UsdPrim prim ) : m_root( root ), m_prim( prim )
+{
+
+}
+
+USDScene::~USDScene()
+{
+}
+
+std::string USDScene::fileName() const
+{
+	return m_root->fileName();
+}
+
+Imath::Box3d USDScene::readBound( double time ) const
+{
+	pxr::TfTokenVector purposes;
+	purposes.push_back( pxr::UsdGeomTokens->default_ );
+	pxr::UsdGeomBBoxCache bboxCache( m_root->getTime( time ), purposes );
+
+	pxr::GfBBox3d bbox = bboxCache.ComputeLocalBound( m_prim );
+
+	return convert( bbox );
+}
+
+ConstDataPtr USDScene::readTransform( double time ) const
+{
+	return new IECore::M44dData( readTransformAsMatrix( time ) );
+}
+
+Imath::M44d USDScene::readTransformAsMatrix( double time ) const
+{
+	pxr::UsdGeomXformable transformable( m_prim );
+	pxr::GfMatrix4d transform;
+	bool reset = false;
+
+	transformable.GetLocalTransformation( &transform, &reset, m_root->getTime( time ) );
+	return convert( transform );
+}
+
+ConstObjectPtr USDScene::readAttribute( const SceneInterface::Name &name, double time ) const
+{
+	return nullptr;
+}
+
+ConstObjectPtr USDScene::readObject( double time ) const
+{
+	return convert( m_prim, m_root->getTime( time ) );
+}
+
+SceneInterface::Name USDScene::name() const
+{
+	return SceneInterface::Name( m_prim.GetName() );
+}
+
+void USDScene::path( SceneInterface::Path &p ) const
+{
+	std::vector<std::string> parts;
+	pxr::SdfPath path = m_prim.GetPath();
+	boost::split( parts, path.GetString(), boost::is_any_of( "/" ) );
+
+	p.reserve( parts.size() );
+
+	for( auto part : parts )
```

Comment:
Should this be `const auto &`? I'm still getting to grips with C++11 but I would assume that `auto part` is going to make a copy?